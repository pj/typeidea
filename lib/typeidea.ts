import * as crypto from 'crypto';
import {compile} from 'ejs';
import * as fs from 'fs';
import * as prettier from 'prettier';
import * as action from './action';

type Hash = string | null;

export function hashAction(
  action: action.Action,
  previous: string | null
) {
  const hash = crypto.createHash('sha256');

  if (previous !== null) {
    hash.update(previous);
  }

  const fieldsToHash = action.fieldsToHash();
  hash.update(fieldsToHash);

  return hash.digest('hex');
}

/**
* Take a list of types and verify the types and generate any empty types.
*/
export function hashActions(
  hashables: Array<action.Action>,
  requireHashes: boolean
): Array<[number, string]> {
  const newHashes: Array<[number, string]> = [];
  let previousHash = null;
  let createHashes = false;
  for (let n = 0; n < hashables.length; n++) {
    let hashable = hashables[n];
    if (createHashes) {
      if (hashable.hash !== null && hashable.hash !== undefined) {
        throw new Error(`Hash entered when there were previous items without hashes index: ${n} action: ${hashable}`);
      }
      previousHash = hashAction(hashable, previousHash);
      newHashes.push([n, previousHash]);
    } else {
      if (hashable.hash === null || hashable.hash === undefined) {
        if (requireHashes) {
          throw new Error(`Unhashed action found at item ${n} action: ${hashable}`);
        }
        previousHash = hashAction(hashable, previousHash);
        newHashes.push([n, previousHash]);
        createHashes = true;
      } else {
        const expectedHash = hashAction(hashable, previousHash);
        if (expectedHash !== hashable.hash) {
          throw new Error(`Invalid hash at item ${n} expected ${expectedHash} got ${hashable.hash} object: ${hashable}`)
        }
        previousHash = hashable.hash;
      }
    }
  }

  return newHashes;
}

export function addHashes(
  unhashedType: action.Action[],
  hashes: Array<[number, string]>,
  hashTo: number | null
) {
  if (hashes.length === 0) {
    return unhashedType;
  }
  const hashed = unhashedType.slice();
  if (hashTo === null) {
    hashTo = unhashedType.length;
  }
  //for (let i = hashes[0][0]; i < hashTo; i++) {
  //  const hash = hashes[i][1];
  //  const action = unhashedType[i];
  //  const newAction = Object.assign(
  //    Object.create(Object.getPrototypeOf(action)),
  //    action
  //  );
  //  newAction.hash = hash
  //  hashed[i] = newAction;
  //}

  for (let i = 0; i < hashTo; i++) {
    const action = hashed[i];
    if (action.hash === null) {
      for (let [idx, hash] of hashes) {
        if (idx === i) {
          const newAction = Object.assign(
            Object.create(Object.getPrototypeOf(action)),
            action
          );
          newAction.hash = hash
          hashed[i] = newAction;
        }
      }
    }
  }

  return hashed;
}
