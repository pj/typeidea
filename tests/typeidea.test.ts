import * as typeidea from '../lib/typeidea';
import * as action from '../lib/action';
import * as generate from '../lib/generate';

it('hashActions generates hashes for types', () => {
  const addField = [
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test',
      'A useful type'
    ),
    new action.AddFieldTypeAction(
      'Adding a field',
      null,
      'Test',
      'test_field',
      'string',
      'A field that helps testing',
      true,
      null
    ),
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test2',
      'Another useful type'
    ),
    new action.AddFieldTypeAction(
      'asdf',
      null,
      'Test2',
      'test_field2',
      'string',
      'A field that helps testing',
      true,
      null
    )
  ];
  const hashes = typeidea.hashActions(addField, false);
  expect(hashes).toHaveLength(4);

  const hashedAddField = typeidea.addHashes(addField, hashes, null);
  expect(hashedAddField).toHaveLength(4);

  for (const hashedAction of hashedAddField) {
    expect(hashedAction.hash).not.toBeNull();
  }
});

it('Changing an action makes hashing invalid', () => {
  const addField = [
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test',
      'a useful type'
    ),
    new action.AddFieldTypeAction(
      'Adding a field',
      null,
      'Test',
      'test_field',
      'string',
      'A field that helps testing',
      true,
      null
    )
  ];

  const hashes = typeidea.hashActions(addField, false);
  const hashedAddField = typeidea.addHashes(addField, hashes, null);
  hashedAddField[0].changeLog = "Don't do this!";
  expect(() => {
    typeidea.hashActions(hashedAddField, true);
  }).toThrow(/Invalid hash at item \d+ .*/);
});

it('Changing a hash makes hashing invalid', () => {
  const addField = [
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test',
      'a useful type'
    ),
    new action.AddFieldTypeAction(
      'Adding a field',
      null,
      'Test',
      'test_field',
      'string',
      'A field that helps testing',
      true,
      null
    )
  ];

  const hashedAddField = typeidea.addHashes(
    addField,
    typeidea.hashActions(addField, false),
    null
  );
  hashedAddField[0].hash = "Don't do this!";
  expect(() => {
    typeidea.hashActions(hashedAddField, true);
  }).toThrow(/Invalid hash at item \d+ .*/);
});

it('Multiple types with type reference', () => {
  const addField = [
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test',
      'A useful type'
    ),
    new action.AddFieldTypeAction(
      'Adding a field',
      null,
      'Test',
      'test_field',
      'string',
      'A field that helps testing',
      true,
      null
    ),
    new action.NewTypeAction(
      'Creating a new type',
      null,
      'Test2',
      'a useful type'
    ),
    new action.ReferenceFieldTypeAction(
      'asdf',
      null,
      'Test2',
      'test_field2',
      'A reference',
      true,
      'Test',
      // @ts-ignore
      null
    )
  ];

  let hashes = typeidea.hashActions(addField, false);
  let hashedAddField = typeidea.addHashes(addField, hashes, 2);

  addField[3].referenceHash = hashedAddField[1].hash;

  // Now that we have the correct hash, generate all
  hashes = typeidea.hashActions(addField, false);
  hashedAddField = typeidea.addHashes(addField, hashes, null);

  const [generatedTypes, generatedServices] = generate.generateDefinitions(
    hashedAddField,
  );
  const typescript = generate.generateTypescript(
    generatedTypes,
  );

  expect(typescript).toMatchSnapshot();
});

const json_snapshot_tests = ([
  ['Add a field', '../tests/test_data/add_field.json', null],
  [
    'Add a field with a default value',
    '../tests/test_data/add_field_with_default_value.json',
    null
  ],
  ['Rename a field', '../tests/test_data/rename_field.json', null],
  ['Make a field optional', '../tests/test_data/optional_field.json', null],
  ['Delete a field', '../tests/test_data/delete_field.json', null],
  ['Type with latest', '../tests/test_data/type_with_latest.json', 3],
  ['Type with GroupAction', '../tests/test_data/type_with_group.json', null],
  [
    'Type with deprecated and dont generate',
    '../tests/test_data/deprecated_and_dont_generated.json',
    null
  ],
] as Array<[string, string, number | null]>);

for (const [name, path, hashTo] of json_snapshot_tests) {
  it(name, () => {
    const addField = action.loadActionLog(path);

    const hashes = typeidea.hashActions(addField, false);
    const hashedAddField = typeidea.addHashes(addField, hashes, hashTo);

    const [generatedTypes, generatedServices] = generate.generateDefinitions(
      hashedAddField
    );
    const typescript = generate.generateTypescript(generatedTypes);

    expect(typescript).toMatchSnapshot();
  });
}

