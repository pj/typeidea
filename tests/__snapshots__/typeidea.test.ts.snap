// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add a field 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d
};

export class Test {
  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\":
        return new Test_V1(message.test_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Add a field with a default value 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c\\";
    this.test_field = test_field || \\"A value\\";
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c
};

export class Test {
  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c\\":
        return new Test_V1(message.test_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Delete a field 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 * 2. Adding another field
 * 3. didn't actually need that
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d
};

/**
 * @param test_field A field that helps testing
 * @param test_field2 A field that helps testing, again
 *
 * @sealed
 */
class Test_V2 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;
  readonly test_field2: string;

  constructor(test_field: string | null, test_field2: string) {
    this.version = 2;
    this.hash =
      \\"4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\";
    this.test_field = test_field;
    this.test_field2 = test_field2 || \\"blah\\";
  }

  static deserialize(message: any): Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V2(message.test_field, message.test_field2);
  }

  static serialize(message: Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V2,
  Test_V2 as Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0
};

/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V3 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 3;
    this.hash =
      \\"d6cc2965f8ee5ebb6de4a255956c869a8c212bcc4d0bccf1dff40a440c2343f7\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V3 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V3(message.test_field);
  }

  static serialize(message: Test_V3): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V3,
  Test_V3 as Test_Hd6cc2965f8ee5ebb6de4a255956c869a8c212bcc4d0bccf1dff40a440c2343f7
};

export class Test {
  static deserialize(message: any): Test_V1 | Test_V2 | Test_V3 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\":
        return new Test_V1(message.test_field);

      case \\"Test_V2\\":
      case \\"Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\":
        return new Test_V2(message.test_field, message.test_field2);

      case \\"Test_V3\\":
      case \\"Test_Hd6cc2965f8ee5ebb6de4a255956c869a8c212bcc4d0bccf1dff40a440c2343f7\\":
        return new Test_V3(message.test_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1 | Test_V2 | Test_V3): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Make a field optional 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 * 2. Oops needed to make it optional
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string;

  constructor(test_field: string) {
    this.version = 1;
    this.hash =
      \\"3b9ac56de579e9a69937f2d1d2c27a4cfbedfd0710682f8ab629b18aac4efe14\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H3b9ac56de579e9a69937f2d1d2c27a4cfbedfd0710682f8ab629b18aac4efe14
};

/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V2 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 2;
    this.hash =
      \\"818de92c4372ef16f0eba77a342f97cd788f18b24d52b54a2870236d82184a0b\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V2(message.test_field);
  }

  static serialize(message: Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V2,
  Test_V2 as Test_H818de92c4372ef16f0eba77a342f97cd788f18b24d52b54a2870236d82184a0b
};

export class Test {
  static deserialize(message: any): Test_V1 | Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H3b9ac56de579e9a69937f2d1d2c27a4cfbedfd0710682f8ab629b18aac4efe14\\":
        return new Test_V1(message.test_field);

      case \\"Test_V2\\":
      case \\"Test_H818de92c4372ef16f0eba77a342f97cd788f18b24d52b54a2870236d82184a0b\\":
        return new Test_V2(message.test_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1 | Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Multiple types with type reference 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * A useful type
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"c68b1dfae49d22c7332dcfcb86145926907e4171e3a7bb1f3861e9e3b0c2bdfb\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_Hc68b1dfae49d22c7332dcfcb86145926907e4171e3a7bb1f3861e9e3b0c2bdfb
};

export class Test {
  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_Hc68b1dfae49d22c7332dcfcb86145926907e4171e3a7bb1f3861e9e3b0c2bdfb\\":
        return new Test_V1(message.test_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

/*
 * Test2
 * a useful type
 * Change Log:
 * 0. Creating a new type
 * 1. asdf
 */
/**
 * @param test_field2 A reference
 *
 * @sealed
 */
class Test2_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field2: Test.h_c68b1dfae49d22c7332dcfcb86145926907e4171e3a7bb1f3861e9e3b0c2bdfb;

  constructor(
    test_field2: Test.h_c68b1dfae49d22c7332dcfcb86145926907e4171e3a7bb1f3861e9e3b0c2bdfb
  ) {
    this.version = 1;
    this.hash =
      \\"d91d0ff7ca6ca8fafe09cc36bf88bf85d5fd3767f68717869c948825843d4aa1\\";
    this.test_field2 = test_field2;
  }

  static deserialize(message: any): Test2_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test2_V1(message.test_field2);
  }

  static serialize(message: Test2_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test2_V1,
  Test2_V1 as Test2_Hd91d0ff7ca6ca8fafe09cc36bf88bf85d5fd3767f68717869c948825843d4aa1
};

export class Test2 {
  static deserialize(message: any): Test2_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test2_V1\\":
      case \\"Test2_Hd91d0ff7ca6ca8fafe09cc36bf88bf85d5fd3767f68717869c948825843d4aa1\\":
        return new Test2_V1(message.test_field2);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test2_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Rename a field 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 * 2. Renaming a field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c\\";
    this.test_field = test_field || \\"A value\\";
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c
};

/**
 * @param renamed_field A field that helps testing
 *
 * @sealed
 */
class Test_V2 {
  readonly version: number;
  readonly hash: string;
  readonly renamed_field: string | null;

  constructor(renamed_field: string | null) {
    this.version = 2;
    this.hash =
      \\"e41f7b607671faf485204483dcf344ef15bbd2856a51485d9876e73178fe07bf\\";
    this.renamed_field = renamed_field || \\"A value\\";
  }

  static deserialize(message: any): Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V2(message.renamed_field);
  }

  static serialize(message: Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V2,
  Test_V2 as Test_He41f7b607671faf485204483dcf344ef15bbd2856a51485d9876e73178fe07bf
};

export class Test {
  static deserialize(message: any): Test_V1 | Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H036cb58d4009b695ef9c0a35e42cfd356c5ca922f3340d79db2364efd0d0da4c\\":
        return new Test_V1(message.test_field);

      case \\"Test_V2\\":
      case \\"Test_He41f7b607671faf485204483dcf344ef15bbd2856a51485d9876e73178fe07bf\\":
        return new Test_V2(message.renamed_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1 | Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Type with deprecated and dont generate 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 * 2. Adding another field
 * 3. Adding yet another field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d
};

/**
 * @param test_field A field that helps testing
 * @param test_field2 A field that helps testing, again
 *
 * @sealed
 */
class Test_V2 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;
  readonly test_field2: string;

  constructor(test_field: string | null, test_field2: string) {
    this.version = 2;
    this.hash =
      \\"4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\";
    this.test_field = test_field;
    this.test_field2 = test_field2 || \\"blah\\";
  }

  static deserialize(message: any): Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V2(message.test_field, message.test_field2);
  }

  static serialize(message: Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V2,
  Test_V2 as Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0
};

/**
 * @param test_field A field that helps testing
 * @param test_field2 A field that helps testing, again
 * @param test_field3 Yet another field for testing
 *
 * @sealed
 */
class Test_V3 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;
  readonly test_field2: string;
  readonly test_field3: number;

  constructor(
    test_field: string | null,
    test_field2: string,
    test_field3: number
  ) {
    this.version = 3;
    this.hash =
      \\"abb52b859471c90b51d0b3103f4ce61cd35664c4e579b49597721df71fd6c655\\";
    this.test_field = test_field;
    this.test_field2 = test_field2 || \\"blah\\";
    this.test_field3 = test_field3;
  }

  static deserialize(message: any): Test_V3 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V3(
      message.test_field,
      message.test_field2,
      message.test_field3
    );
  }

  static serialize(message: Test_V3): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V3,
  Test_V3 as Test_Habb52b859471c90b51d0b3103f4ce61cd35664c4e579b49597721df71fd6c655
};

export class Test {
  static deserialize(message: any): Test_V1 | Test_V2 | Test_V3 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\":
        return new Test_V1(message.test_field);

      case \\"Test_V2\\":
      case \\"Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\":
        return new Test_V2(message.test_field, message.test_field2);

      case \\"Test_V3\\":
      case \\"Test_Habb52b859471c90b51d0b3103f4ce61cd35664c4e579b49597721df71fd6c655\\":
        return new Test_V3(
          message.test_field,
          message.test_field2,
          message.test_field3
        );

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1 | Test_V2 | Test_V3): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`Type with latest 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * Test
 * Test
 * Change Log:
 * 0. Creating a new type
 * 1. Adding a field
 * 2. Adding another field
 * 3. Adding yet another field
 */
/**
 * @param test_field A field that helps testing
 *
 * @sealed
 */
class Test_V1 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;

  constructor(test_field: string | null) {
    this.version = 1;
    this.hash =
      \\"56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\";
    this.test_field = test_field;
  }

  static deserialize(message: any): Test_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V1(message.test_field);
  }

  static serialize(message: Test_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V1,
  Test_V1 as Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d
};

/**
 * @param test_field A field that helps testing
 * @param test_field2 A field that helps testing, again
 *
 * @sealed
 */
class Test_V2 {
  readonly version: number;
  readonly hash: string;
  readonly test_field: string | null;
  readonly test_field2: string;

  constructor(test_field: string | null, test_field2: string) {
    this.version = 2;
    this.hash =
      \\"4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\";
    this.test_field = test_field;
    this.test_field2 = test_field2 || \\"blah\\";
  }

  static deserialize(message: any): Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new Test_V2(message.test_field, message.test_field2);
  }

  static serialize(message: Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  Test_V2,
  Test_V2 as Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0
};

export class Test {
  static deserialize(message: any): Test_V1 | Test_V2 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"Test_V1\\":
      case \\"Test_H56fe5eaa9656364c83c84208b60dd08e81bc05eb37816cc21eb93618af57150d\\":
        return new Test_V1(message.test_field);

      case \\"Test_V2\\":
      case \\"Test_H4ce3d4cfba96f0cdcce4bf1d2812eea0c5b1cf529ac93763f6f33d0600a652f0\\":
        return new Test_V2(message.test_field, message.test_field2);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: Test_V1 | Test_V2): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;
